File: client/src/services/phantombuster.js
/**
 * PhantomBuster API Service
 * Frontend service for interacting with PhantomBuster through the backend API
 */

import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://127.0.0.1:8000';

/**
 * Get list of all PhantomBuster agents
 */
export async function getAgents() {
  const response = await axios.get(`${API_BASE_URL}/api/phantombuster/agents`);
  return response.data;
}

/**
 * Get status and details of a specific agent
 * @param {string} agentId - The ID of the agent
 */
export async function getAgentStatus(agentId) {
  const response = await axios.get(`${API_BASE_URL}/api/phantombuster/agents/${agentId}`);
  return response.data;
}

/**
 * Get output/results from a specific agent
 * @param {string} agentId - The ID of the agent
 * @param {string} mode - Either "most-recent" or "all"
 */
export async function getAgentOutput(agentId, mode = 'most-recent') {
  const response = await axios.get(
    `${API_BASE_URL}/api/phantombuster/agents/${agentId}/output`,
    { params: { mode } }
  );
  return response.data;
}

/**
 * Launch an agent to start scraping
 * @param {string} agentId - The ID of the agent
 * @param {Object} argument - Optional parameters to pass to the agent
 */
export async function launchAgent(agentId, argument = null) {
  const response = await axios.post(
    `${API_BASE_URL}/api/phantombuster/agents/launch`,
    { agent_id: agentId, argument }
  );
  return response.data;
}

/**
 * Get list of all result containers
 */
export async function getContainers() {
  const response = await axios.get(`${API_BASE_URL}/api/phantombuster/containers`);
  return response.data;
}

/**
 * Get data from a specific result container
 * @param {string} containerId - The ID of the container
 */
export async function getContainerData(containerId) {
  const response = await axios.get(`${API_BASE_URL}/api/phantombuster/containers/${containerId}`);
  return response.data;
}

/**
 * PhantomBuster service object with all methods
 */
const phantombusterService = {
  getAgents,
  getAgentStatus,
  getAgentOutput,
  launchAgent,
  getContainers,
  getContainerData
};

export default phantombusterService;

File: server/app/services/phantombuster.py
"""
PhantomBuster API Client
Provides methods to interact with PhantomBuster API for LinkedIn scraping
"""

import os
from typing import Dict, List, Optional
import requests
from dotenv import load_dotenv

load_dotenv()


class PhantomBusterClient:
    """Client for interacting with PhantomBuster API"""

    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize the PhantomBuster client

        Args:
            api_key: Your PhantomBuster API key (defaults to env var)
        """
        self.api_key = api_key or os.getenv("PHANTOMBUSTER_API_KEY")
        if not self.api_key:
            raise ValueError("PhantomBuster API key is required")

        self.base_url = "https://api.phantombuster.com/api/v2"
        self.headers = {
            "X-Phantombuster-Key": self.api_key,
            "Content-Type": "application/json"
        }

    def get_agents(self) -> List[Dict]:
        """Get list of all your agents (phantoms)"""
        response = requests.get(
            f"{self.base_url}/agents/fetch-all",
            headers=self.headers,
            timeout=30
        )
        response.raise_for_status()
        return response.json()

    def get_agent_output(self, agent_id: str, mode: str = "most-recent") -> Dict:
        """
        Get the output/results from a specific agent

        Args:
            agent_id: The ID of your agent/phantom
            mode: Either "most-recent" or "all"
        """
        response = requests.get(
            f"{self.base_url}/agents/fetch-output",
            headers=self.headers,
            params={"id": agent_id, "mode": mode},
            timeout=30
        )
        response.raise_for_status()
        return response.json()

    def launch_agent(self, agent_id: str, argument: Optional[Dict] = None) -> Dict:
        """
        Launch an agent to start scraping

        Args:
            agent_id: The ID of your agent/phantom
            argument: Optional parameters to pass to the agent
        """
        payload = {"id": agent_id}
        if argument:
            payload["argument"] = argument

        response = requests.post(
            f"{self.base_url}/agents/launch",
            headers=self.headers,
            json=payload,
            timeout=30
        )
        response.raise_for_status()
        return response.json()

    def get_containers(self) -> List[Dict]:
        """Get list of all result containers"""
        response = requests.get(
            f"{self.base_url}/containers/fetch-all",
            headers=self.headers,
            timeout=30
        )
        response.raise_for_status()
        return response.json()

    def get_container_data(self, container_id: str) -> Dict:
        """
        Get data from a specific result container
        This returns container metadata, not the actual data

        Args:
            container_id: The ID of the container
        """
        response = requests.get(
            f"{self.base_url}/containers/fetch",
            headers=self.headers,
            params={"id": container_id},
            timeout=30
        )
        response.raise_for_status()
        return response.json()

    def get_agent_result_object(self, agent_id: str) -> Dict:
        """
        Get the result object (CSV/JSON URLs) from an agent

        Args:
            agent_id: The ID of your agent/phantom
        """
        response = requests.get(
            f"{self.base_url}/agents/fetch-result-object",
            headers=self.headers,
            params={"id": agent_id},
            timeout=30
        )
        response.raise_for_status()
        return response.json()

    def get_agent_status(self, agent_id: str) -> Dict:
        """
        Get the current status of an agent

        Args:
            agent_id: The ID of your agent/phantom
        """
        response = requests.get(
            f"{self.base_url}/agents/fetch",
            headers=self.headers,
            params={"id": agent_id},
            timeout=30
        )
        response.raise_for_status()
        return response.json()


# Singleton instance for dependency injection
_client_instance: Optional[PhantomBusterClient] = None


def get_phantombuster_client() -> PhantomBusterClient:
    """Get or create the PhantomBuster client singleton"""
    global _client_instance
    if _client_instance is None:
        _client_instance = PhantomBusterClient()
    return _client_instance
